### DISCUSSION TITLE
Represent Variable Packet Structure in XTCE

### DISCUSSION THREAD

**Discussion by msarrel-work** [yamcs/yamcs#959] [Q&A]
I'm not sure how to represent the structure of some of our packets in XTCE.  The main problem I have is how to specify a parameter whose length is a function of the first fixed portion of its content.

Those packets have a variable structure in the data portion.  They contain a list of telemetry channels with the following constraints.

1) Each telemetry channel in the list contains a channel identifier (fixed size), and a channel value (variable size, a function of the channel identifier). The size is found by a look-up table on the ground.

2) Each packet may contain a list that contains a few dozen telemetry channels.

3) The telemetry channels in the list may appear in any order.

4) The telemetry channels are selected by flight software from a possible set of several thousand channel identifiers.

5) There is no way the ground can know in advance what telemetry channels, and in which order, will be in a given packet.  That is determined when flight software creates the packet on the spacecraft.

The best I can determine is to use this portion of the XTCE schema:

- SequenceContainerType
- EntryList:  EntryListType
- ParameterRefEntry:  ParameterRefEntryType
- LocationInContainerInBits:  LocationInContainerInBitsType
- DiscreetLookupList: DiscreetLookupListType

Then, within that, I'm thinking or one of these two options:

- DynamicValue: DynamicValueType
- DiscreetLookup: DiscreetLookupType

This might involve the use of custom algorithm in the DiscreetLookupList and that would be OK.

The LocationInContainerBits wants to specify the start of the entry relative to the end of the previous entry.  If they're back-to-back, the offset is zero.

> The start bit 0 position for each container is local to the container, but does include space occupied by inherited containers.  When a container is "included", as opposed to inherited, then the interpreting implementation takes into account the start bit position of the referring container when finally assembling the start bits for the post-processed entry content.  The default start bit for any entry is 0 bits from the previous entry, making the content contiguous when this element is not used.

What I really need is a way to dynamically specify the length of an entry.  Instead, this seems to specify a gap between two entries.

Maybe I could pretend that the entries only consist of the channel identifiers and the "gaps" are the channel values.  But, then I'm not sure how I'd get at the channel values themselves.  XTCE would consider that they contain nothing.

Another option would be to tell XTCE nothing about the structure of the data portion of these packets and simply use a YAMCS custom function that we write to parse the data portion.  Is that possible, and if so, can you point me to the section portion of the documetation?

**Comment by xpromache** [comment]
Maybe you need the IndirectParamRefEntry, see this discussion:

https://groups.google.com/g/yamcs/c/x4JwxtO3VM8/m/hGRPmwhdAgAJ

Currently it doesn't work in Yamcs but there is a pull request which will be merged soon.

**Reply by msarrel-work** [reply]
Thank you.  This example looks very much like my problem.  Do you have an estimate when #932 might be merged?

**Reply by xpromache** [reply]
in a couple of hours

**Reply by msarrel-work** [reply]
@xpromache Thank you for the information, it was very helpful.  I made it work partially.  I think there is still a bug in YAMCS that prevents complete success.

Our packets are slightly different from José's in the link you provided.  Our packets do not indicate the number of parameters.  We leave it to the ground to count the parameters.  Our packets contain a series of pairs of MeaurementIDs and values.  The MeaurementID is sixteen bits, and the length of the value is deduced from the MeasurementID.  For this simple example, I fixed the number of parameters per packet first at one, and then at three.  YAMCS decodes the packets correctly when the number is one, but incorrectly when the number is three.

When the number of parameters is fixed at one, the packets are decoded correctly.

![image](https://github.com/user-attachments/assets/2088eae4-d52b-4d0a-866e-99b63e4d355e)
![image](https://github.com/user-attachments/assets/a1c35a31-31d7-466e-8b9a-ad117de5685e)
![image](https://github.com/user-attachments/assets/e32da51e-9877-4bde-9fe5-d8bdbfd250c6)

YAMCS correctly determines the length of the value for each MeasurementID.  MeasurementID 15 is sixteen bits, MeasurementID 13 is eight bits and MeaurementID 11 is thirty-two bits.

When the number of parameters per packet is fixed at three, the packet is decoded incorrectly.

![image](https://github.com/user-attachments/assets/add234ee-b430-42fb-b572-33ae8f5ebcf6)

Here is seems that YAMCS assumes that it has three parameters, all with the MeasurementID of 15 and length of sixteen bits.

Can you help?

Best Regards,

Marc

**Comment by xpromache** [comment]
This was the original bug which i thought we solved with the release 10.5.5.

Can you share your MDB definition?

**Reply by msarrel-work** [reply]
Do you mean release 5.10.5?  I'm using the YAMCS QuickStart repo, which is still on 5.10.4.  I'll try the YAMCS full repo, which looks to be on 5.10.6.

**Reply by xpromache** [reply]
Actually it should have been solved in 5.10.4.
But to make sure just update the yamcs version in the pom.xml of the quickstart project: just set it to the latest version 5.10.6

**Reply by msarrel-work** [reply]
I upgraded the pom.xml to 5.10.6, but the problem persists.  It is difficult for me to share the MDB version.  I'll have to wait until next week to get approval.  Can you send me an MDB that works for this situation and I'll try to adapt it?

**Reply by xpromache** [reply]
You are right, this thing is still broken, sorry for that. 
I'm going to fix it right away.

**Reply by xpromache** [reply]
I pushed a fix to the master branch.
If you compile yamcs locally (mvn install -DskipTests in the main yamcs repo) then you can set the version "5.10.7-SNAPSHOT" in the pom.xml of the quickstart.

If you confirm it works, I can make a release sometime in the next days.

**Reply by msarrel-work** [reply]
I must be doing something wrong.  In local yamcs repo, I did the following:

```
mvn clean install -DskipTests
cd yamcs-web/src/main/webapp
npm install
npm run build
cd -
```

Then, in the QuickStart repo, I changed `pom.xml` like this:

```
    <yamcsVersion>5.10.7-SNAPSHOT</yamcsVersion>
```

But, now I cannot connect to `localhost:8090`.  In the web browser I see `404 Not Found`.  And, in the yamcs log, I see:

```
13:52:10.629 _global [65] AngularHandler Responding '404 Not Found': /Users/msarrel/myproject/target/yamcs/cache/yamcs-web/index.html
13:52:10.629 _global [65] HttpRequestHandler c3725f82 GET / 404
```

**Reply by xpromache** [reply]
I forgot about npm. 
Please do a mvn install -DkipTests after running npm, as it will package the web stuff in the yamcs-web jar.

**Reply by msarrel-work** [reply]
Success!

Thank you.  It works now:

![image](https://github.com/user-attachments/assets/8e1999f9-16ce-4905-9448-d78382843cca)
